### PR TITLE
implement `groups` for conv1d and conv2d

### PR DESCRIPTION
**Pull request by Jimexist**
## Proposed changes

- related to and partly fix-es #100 

## Checklist

Put an `x` in the boxes that apply.

- [ ] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

**Comment by jagrit06** [comment]
Thanks for putting this together! 
I'm putting together an update to the convolutions that won't have groups handled in the backend right away - so this would be a good way to add that functionality till that specialization is ready!

I'm just dropping a comment here to give you a heads up so that you can be on the look out for any updates and PRs to conv in the meantime

File: python/mlx/nn/layers/convolution.py [hunk 1]
```diff
@@ -34,6 +37,8 @@ def __init__(
         kernel_size: int,
         stride: int = 1,
         padding: int = 0,
+        dilation: int = 1,
+        groups: int = 1,
         bias: bool = True,
     ):
         super().__init__()
```
Comment (tic-top) [inline python/mlx/nn/layers/convolution.py:44]: Why the shape of the weight doesn't change? 
I think the shape should be (out_channel, kernel_size, in_channels/groups) and (in_channel%groups ==0).

File: python/mlx/nn/layers/convolution.py [hunk 2]
```diff
@@ -92,13 +106,15 @@ def __init__(
...
 
-        kernel_size, stride, padding = map(
+        kernel_size, stride, padding, dilation = map(
             lambda x: (x, x) if isinstance(x, int) else x,
-            (kernel_size, stride, padding),
+            (kernel_size, stride, padding, dilation),
         )
         scale = math.sqrt(1 / (in_channels * kernel_size[0] * kernel_size[1]))
         self.weight = mx.random.uniform(
```
Comment (tic-top) [inline python/mlx/nn/layers/convolution.py:120]: Same question here, in_channel/groups

File: mlx/ops.cpp [hunk 3]
```diff
@@ -2767,6 +2768,22 @@ inline void run_conv_checks(const array& in, const array& wt, int n_dim) {
         << " input: " << in.shape() << " and weight: " << wt.shape();
```
Comment (tic-top) [inline mlx/ops.cpp:2768]: If the group>1, then the last dimension of weight * group == the last dimension of input shape

**Comment by tic-top** [comment]
After I fixed the bugs, I had a try on depthwise conv1d, very slow. 
I can't wait to see convolution update!

**Comment by a1eaiactaest** [comment]
@Jimexist any updates on this?

**Comment by Jimexist** [comment]
> @Jimexist any updates on this?

i paused working on this due to @jagrit06's message. not sure if our paths will cross and duplicate work

**Comment by awni** [comment]
What's the status of this PR? I think we could move it forward as no-one else is working on groups right now? Wdyt? 

Otherwise let's close it so someone else can pick this up.

**Comment by Jimexist** [comment]
> What's the status of this PR? I think we could move it forward as no-one else is working on groups right now? Wdyt?
> 
> Otherwise let's close it so someone else can pick this up.

in that case let me see if i can move on and implement this, also rebasing after:
- https://github.com/ml-explore/mlx/pull/766

**Comment by Stealeristaken** [comment]
Is that PR still active? I would like to work on it.

**Comment by Jimexist** [comment]
> Is that PR still active? I would like to work on it.

@Stealeristaken feel free to take over since i am bit busy and engaged in March

**Pull request closed without merging**